🐛 BUG: `wrangler dev` sends request to hostname of custom zone, instead of full domain

### What version of `Wrangler` are you using?

2.1.11

### What operating system are you using?

Linux

### Describe the Bug

I've got my domain `example.com`, and a Worker deployed at `worker.example.com`. The second-level domain hosts a website, which is totally unrelated to what my Worker does.

I use a grey-clouded DNS record pointing to my server for `example.com`, and an orange-clouded DNS record pointing to `100::` for `worker.example.com`.

When trying to test the Worker using `wrangler dev`, an error message pops up with the content of a typical 404 reply from my `example.com` website. Publishing the Worker works fine and interacting with it once published is also fine. Changing the Workers route to the default `workers.dev` domain also makes the error disappear.

Though not being able to test my Worker this way is annoying, my main concern is that I don't understand why on Earth a request to `example.com` is done in the process.

Sure, here it is. I removed most of the HTML content to make it more brief.

```
 ⛅️ wrangler 2.1.11 
--------------------
Your worker has access to the following bindings:
⬣ Listening at http://0.0.0.0:8787
- http://127.0.0.1:8787
- http://192.168.1.118:8787
- http://192.168.97.1:8787
- http://192.168.143.1:8787
✘ [ERROR] Error while creating remote dev session: ParseError: Unexpected token <

      at parseJSON
  (/home/user/project/node_modules/wrangler/wrangler-dist/cli.js:136608:11)
      at createPreviewSession
  (/home/user/project/node_modules/wrangler/wrangler-dist/cli.js:143712:51)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async start
  (/home/user/project/node_modules/wrangler/wrangler-dist/cli.js:144345:9) {
    text: 'Unexpected token <',
    notes: [],
    location: {
      lineText: '<!DOCTYPE html>',
      line: 1,
      column: -1,
      file: undefined,
      fileText: '<!DOCTYPE html>\n' +
        '<html>\n' +
        '<head>\n' +
        '<meta charset="utf-8">\n'
        
    A whole lot of HTML...
        
    ... 1939 more characters
    },
    kind: 'error'
  }

```

--- a/.changeset/happy-grapes-check.md
+++ b/.changeset/happy-grapes-check.md
@@ -1,0 +1,5 @@
+---
+"wrangler": minor
+---
+
+Previously, wrangler dev would not work if the root of your zone wasn't behind Cloudflare. This behaviour has changed so that now only the route which your Worker is exposed on has to be behind Cloudflare.

--- a/packages/wrangler/src/create-worker-preview.ts
+++ b/packages/wrangler/src/create-worker-preview.ts
@@ -109,6 +109,14 @@ function randomId(): string {
 	});
 }
 
+// URLs are often relative to the zone. Sometimes the base zone
+// will be grey-clouded, and so the host must be swapped out for
+// the worker route host, which is more likely to be orange-clouded
+function switchHost(originalUrl: string, host?: string): URL {
+	const url = new URL(originalUrl);
+	url.hostname = host ?? url.hostname;
+	return url;
+}
 /**
  * Generates a preview session token.
  */
@@ -122,12 +130,14 @@ export async function createPreviewSession(
 		? `/zones/${ctx.zone}/workers/edge-preview`
 		: `/accounts/${accountId}/workers/subdomain/edge-preview`;
 
-	const { exchange_url } = await fetchResult<{ exchange_url: string }>(
+	let { exchange_url } = await fetchResult<{ exchange_url: string }>(
 		initUrl,
 		undefined,
 		undefined,
 		abortSignal
 	);
+
+	exchange_url = switchHost(exchange_url, ctx.host).toString();
 
 	logger.debug(`-- START EXCHANGE API REQUEST: GET ${exchange_url}`);
 	logger.debug("-- END EXCHANGE API REQUEST");
@@ -147,16 +157,15 @@ export async function createPreviewSession(
 		token: string;
 		prewarm: string;
 	}>(bodyText);
-
-	const { host } = new URL(inspector_websocket);
-	const query = `cf_workers_preview_token=${token}`;
+	const inspector = new URL(inspector_websocket);
+	inspector.searchParams.append("cf_workers_preview_token", token);
 
 	return {
 		id: randomId(),
 		value: token,
-		host,
-		inspectorUrl: new URL(`${inspector_websocket}?${query}`),
-		prewarmUrl: new URL(prewarm),
+		host: ctx.host ?? inspector.host,
+		inspectorUrl: switchHost(inspector.href, ctx.host),
+		prewarmUrl: switchHost(prewarm, ctx.host),
 	};
 }
 

--- a/packages/wrangler/src/create-worker-preview.ts
+++ b/packages/wrangler/src/create-worker-preview.ts
@@ -130,18 +130,20 @@ export async function createPreviewSession(
 		? `/zones/${ctx.zone}/workers/edge-preview`
 		: `/accounts/${accountId}/workers/subdomain/edge-preview`;
 
-	let { exchange_url } = await fetchResult<{ exchange_url: string }>(
+	const { exchange_url } = await fetchResult<{ exchange_url: string }>(
 		initUrl,
 		undefined,
 		undefined,
 		abortSignal
 	);
 
-	exchange_url = switchHost(exchange_url, ctx.host).toString();
+	const switchedExchangeUrl = switchHost(exchange_url, ctx.host).toString();
 
-	logger.debug(`-- START EXCHANGE API REQUEST: GET ${exchange_url}`);
+	logger.debug(`-- START EXCHANGE API REQUEST: GET ${switchedExchangeUrl}`);
 	logger.debug("-- END EXCHANGE API REQUEST");
-	const exchangeResponse = await fetch(exchange_url, { signal: abortSignal });
+	const exchangeResponse = await fetch(switchedExchangeUrl, {
+		signal: abortSignal,
+	});
 	const bodyText = await exchangeResponse.text();
 	logger.debug(
 		"-- START EXCHANGE API RESPONSE:",

